pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AbstractJavaNode.java

```diff
@@ -89,19 +99,161 @@ public Scope getScope() {
+     *
+     * @param child Child to add
+     * @param index Index the child should have in the parent
+     */
+    // visible to parser only
+    void insertChild(AbstractJavaNode child, int index, boolean overwrite) {
+        // Allow to insert a child without overwriting
+        // If the child is null, it is replaced. If it is not null, children are shifted
+        if (index <= children.length) {
+            if (overwrite || children[index] == null) {
+                children[index] = child;
+            } else {
+                children = ArrayUtils.insert(index, children, child);
+            }
+        }
+        child.jjtSetChildIndex(index);
+        child.jjtSetParent(this);
+
+        updateChildrenIndices(index);
+
+        // The text coordinates of this node will be enlarged with those of the child
+
+        enlargeOnInsert(index, child);
+    }
+
+    private void enlargeOnInsert(int index, AbstractJavaNode child) {
+        if (index == 0) {
+            enlargeLeft(child);
+        }
+        if (index == children.length - 1) {
+            enlargeRight(child);
+        }
+    }
+
+    private void enlargeLeft() {
+        if (jjtGetNumChildren() > 0) {
+            enlargeLeft((AbstractJavaNode) jjtGetChild(0));
+        }
+    }
+
+    private void enlargeLeft(AbstractJavaNode child) {
+        if (this.beginLine > child.beginLine) {
+            this.firstToken = child.firstToken;
+            this.beginLine = child.beginLine;
+            this.beginColumn = child.beginColumn;
+        } else if (this.beginLine == child.beginLine
+            && this.beginColumn > child.beginColumn) {
+            this.firstToken = child.firstToken;
+            this.beginColumn = child.beginColumn;
+        }
+    }
+
+    private void enlargeRight(AbstractJavaNode child) {
+        if (this.endLine < child.endLine) {
+            this.lastToken = child.lastToken;
+            this.endLine = child.endLine;
+            this.endColumn = child.endColumn;
+        } else if (this.endLine == child.endLine
+            && this.endColumn < child.endColumn) {
+            this.lastToken = child.lastToken;
+            this.endColumn = child.endColumn;
+        }
+    }
+
+    /**
+     * Updates the {@link #jjtGetChildIndex()} of the children with their
+     * real position, starting at [startIndex].
+     */
+    private void updateChildrenIndices(int startIndex) {
+        if (startIndex < 0) {
+            startIndex = 0;
+        }
+        for (int j = startIndex; j < jjtGetNumChildren(); j++) {
+            children[j].jjtSetChildIndex(j); // shift the children to the right
+            children[j].jjtSetParent(this);
+        }
+    }
+
+
+    /**
+     * Shifts the begin and end columns of this node by the given offsets.
+     */
+    void shiftColumns(int beginShift, int endShift) {
+        this.beginColumn += beginShift;
+        this.endColumn += endShift;
+
+        // TODO change the tokens. We need to link index
```

Comment:
this is very important if we are to rewrite the AST… without this, stuff like autofixes are impossible, as all tokens (and in particular special tokens!) are unavailable and therefore lost

Reply:
Yes it's important. Unless we doubly link tokens to their preceding token as well, we can't really have a general-purpose implementation though. We *could* hack it for this very restricted use case

We should probably stop storing the begin and end positions separately from the tokens. A node's boundaries are necessarily on token boundaries, so there could be a mismatch which makes the whole implementation very fragile (and the fact all setters are available means that *anyone* can put the node into an inconsistent state)... So we could consider making getBeginLine delegate to `firstToken.getLine()` and remove setters for lines/columns. We still would need to doubly link the tokens I think

Reply:
I agree on delegating.

I'm not sure the double linking is required… it would be helpful, true, but also place constraints on all underlying parsers (JavaCC / Antlr) to support it.

Reply:
We can work around the double linking for the time being. But operations like moving the end of a node one token to the left will be inefficient as we have to find the new end token by iterating left to right from the first token, instead of right to left from the end. For the use case of shrinking the bounds of an ambiguous name, that isn't too bad as they don't have many tokens. But a general purpose implementation would need to be more efficient.

Reply:
I've pushed a workaround